pytorch_lightning/trainer/trainer.py

```diff
@@ -436,6 +436,7 @@ def fit(
         self.accelerator.connect(model)
         self.accelerator.setup_environment()
         self.call_setup_hook(model)  # allow user to setup lightning_module in accelerator environment
+        self.call_model_parallel_hook(model)  # allow user to setup in model parallel environment
```

Comment:
why does it have to be a new hook. could it be combined with the setup hook?

Reply:
I honestly don't know which is better, so please advise!

I added the hook because it makes it clearer to the user that this is specific to model parallelism; in this case model parallel plugins (Fully Sharded/DeepSpeed) will wrap this hook in a specific context allowing them to do model parallel as layers are initialised. This could also be done in the setup hook obviously, but makes it more elusive imo.

The other thing is there is some additional branching logic around the call. For DeepSpeed and FSDP we call it once within a training/testing/validation run even if there are a fit/test combo; this is because we want to instantiate the layers once in the model parallel context, and never again in the same session!

Reply:
-1 to merging them. The LightningModule/DataModulel setup hook isn't related at all with the `model_parallel_setup` hook. The usecase is different and the later doesn't get a `stage` argument.

> we call it once within a training/testing/validation run even if there are a fit/test combo

Although this is also the case for the other setups

Reply:
@carmocca just to be clear, the `setup` hook is called multiple times just with different stages right? Whereas the model parallel setup hook is only called once within training or testing, and is never called again, but this depends on the plugin itself via the property: https://github.com/PyTorchLightning/pytorch-lightning/pull/6679/files#diff-667a2513c158c2b01d73b479e4dea75587d8531f9cd286891d34570a2fd145dcR1082

Reply:
if `setup` is called with no stage argument, it won't be called again regardless of training/testing. If it's called with a stage, then it won't be called again for that stage

Reply:
@SeanNaren The docs explain that user can build their layers inside this hook.
Until now, if the user wanted to delay model construction, they did that in the setup hook, and the setup hook gets called on all accelerators/plugins, which made the code agnostic to that. Or is this not correct?

So if we want to avoid duplication, what is your advice? Let the user call setup() in the model_parallel_setup() hook?

Reply:
Thanks @awaelchli, that's the issue I was trying to think through as well. I could imagine two version of the docs:

Obviously I'll add details explaining what things mean:

```
To build your models faster, or to fit a model that doesn't fit on a single machine but will fit on the world memory, you have to define your layers in the `setup` function as below.

class MyModel(pl.LightningModule):
    def setup(self, trainer, stage):
         self.linear = torch.nn.Linear(5, 5)

The setup function will be wrapped within a context manager that shards the modules as they are made. 
```

or

```
To build your models faster, or to fit a model that doesn't fit on a single machine but will fit on the world memory, you have to define your layers in the `configure_parallel_layers` function as below.

class MyModel(pl.LightningModule):
    def configure_parallel_layers(self, stage):
         self.linear = torch.nn.Linear(5, 5)

The setup function will be wrapped within a context manager that shards the modules as they are made.

If you've used the setup function to setup layers as suggested in the docs previous, then you can call the setup function directly in the code:

class MyModel(pl.LightningModule):
    def configure_parallel_layers(self, stage):
         self.setup(stage)
```

Which would you prefer @carmocca @awaelchli @tchaton

Reply:
both are interesting ways, It would be most important that I can create all with a single method if no space limitations and partial if there are some restrictions...
just to not diverse too much from mentioned docs I would slightly prefer calling setup inside parallel